keps/sig-storage/3751-volume-attributes-class/README.md

```diff
@@ -439,31 +439,28 @@ Operation metrics from [csiOperationsLatencyMetric](https://github.com/kubernete
+The **vac_finalizer_controller** sets/removes the finalizer on all VolumeAttributesClasses. This prevents users from deleting a VolumeAttributesClass that's used by a PVC. The vac_finalizer_controller will maintain a cache of a map of VolumeAttributesClass to the list of PVCs that are using the VolumeAttributesClass.
 
 There are a few conditions that will trigger add/remove pvc finalizers in the VolumeAttributesClass:
 
-
 1. PVC created with a VolumeAttributesClass
-
-    The **VACObjectInUseProtection admission controller**:
-    * Check if the VolumeAttributesClass exists. If not, the PVC will enter the INPROGRESS state because we do not want to impose ordering on object creation
+    The **vac_finalizer_controller**:
+    * If the VolumeAttributesClassName is nil or empty, the VolumeAttributesClass will not be added to the vac_finalizer_controller cache
+    * Check if the VolumeAttributesClass exists. If not, the VolumeAttributesClass will not be added to the vac_finalizer_controller cache
     * Check if this VolumeAttributesClass already has a protection finalizer
     * Add the finalizer to the VolumeAttributesClass if there is none 
 2. PVC created with a VolumeAttributesClass being deleted
-    The **VACObjectInUseProtection admission controller**:
+    The **vac_finalizer_controller**:
     * Check VolumeAttributesClass is being deleted and PVC creation failed
 3. PVC updated to a different VolumeAttributesClass
-    * The **VACProtectionController** will remove PVC finalizer in previous VolumeAttributesClass if after listing all the PVCs and confirmed that this PVC is the last one that is consuming the previous VolumeAttributesClass
-
-    The **VACObjectInUseProtection admission controller**:
+    * The **vac_finalizer_controller** will remove finalizer in the VolumeAttributesClass only if after listing all the PVCs/PVs and confirm that this PVC/PV is the last one that is consuming the VolumeAttributesClass in the vac_finalizer_controller cache
     * Check if the new VolumeAttributesClass already has a protection finalizer
     * Add the finalizer to the new VolumeAttributesClass if there is none 
 4. PVC updated to a different VolumeAttributesClass that is being deleted
-    The **VACObjectInUseProtection admission controller**:
+    The **vac_finalizer_controller**:
     * Check VolumeAttributesClass is being deleted and PVC update failed
 5. PVC has a VolumeAttributesClass and this PVC is deleted
-    * The **VACProtectionController** will remove finalizer in the VolumeAttributesClass only if after listing all the PVCs and confirm that this PVC is the last one that is consuming the VolumeAttributesClass
+    * The **vac_finalizer_controller** will remove finalizer in the VolumeAttributesClass only if after listing all the PVCs/PVs and confirm that this PVC/PV is the last one that is consuming the VolumeAttributesClass in the vac_finalizer_controller cache  
```

Comment:
Do you want to mention that the vac_finalizer_controller will use **only** informers and therefore it may remove the finalizer too early? For example, when a new PVC that needs the VAC has been created, but the informers do not know it yet.

It would be nice to have a paragraph below (around line 481) to summarize why a controller only and why informers only is enough and emphasize few examples where it's imperfect:

* when user creates a PVC and it waits for a VAC to get created and then creates the VAC and immediately deletes it. The provisioner will likely start to provision volumes, but vac_finalizer_controller may not have time to  to add the finalizer to the VAC.
* When user deletes a VAC that looks unused, but a PVC without vac_finalizer_controller noticing it

Reply:
> when user creates a PVC and it waits for a VAC to get created and then creates the VAC and immediately deletes it. The provisioner will likely start to provision volumes, but vac_finalizer_controller may not have time to to add the finalizer to the VAC.

Only initializer can prevent this via adding a finalizer to the VAC before the VAC is stored in etcd. For example, extending the existing admission plugin StorageObjectInUseProtection makes it work for VACs as well.

```
func (c *storageProtectionPlugin) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
	switch a.GetResource().GroupResource() {
	case pvResource:
		return c.admitPV(a)
	case pvcResource:
		return c.admitPVC(a)
	case vacResource:
		return c.admitPVC(a)

	default:
		return nil
	}
}
```

Or introduce a new admission plugin for VACs.

Reply:
> When user deletes a VAC that looks unused, but a PVC without vac_finalizer_controller noticing it

sorry, I cannot understand the problem statement. There's a draft implementation. cc @jsafrane https://github.com/kubernetes/kubernetes/pull/123549.

Reply:
WDYT @sunnylovestiramisu @msau42 @xing-yang

Reply:
ACK, will add more details to the race conditions